### PR TITLE
add some cmake errors to prevent linking against GPL libraries

### PR DESCRIPTION
**Pull request by karlnapf**

### CODE REVIEW

File: src/shogun/CMakeLists.txt [hunk 1]
```diff
@@ -360,7 +361,8 @@ SHOGUN_DEPENDENCIES(
 SHOGUN_DEPENDENCIES(
   LIBRARY LibLZMA
   SCOPE PRIVATE
-  CONFIG_FLAG USE_LZMA)
+  CONFIG_FLAG USE_LZMA
+  GPL_ONLY)
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: because liblzma is gpl?

File: src/shogun/CMakeLists.txt [hunk 2]
```diff
@@ -370,7 +372,8 @@ SHOGUN_DEPENDENCIES(
 SHOGUN_DEPENDENCIES(
   LIBRARY LZO
   SCOPE PRIVATE
-  CONFIG_FLAG USE_LZO)
+  CONFIG_FLAG USE_LZO
+  GPL_ONLY)
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: because liblzo is gpl?

File: src/shogun/CMakeLists.txt [hunk 3]
```diff
@@ -400,18 +403,22 @@ ELSE()
   SHOGUN_DEPENDENCIES(
     LIBRARY NLopt
     SCOPE PRIVATE
-    CONFIG_FLAG HAVE_NLOPT)
+    CONFIG_FLAG HAVE_NLOPT
+    GPL_ONLY)
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: because nlopt is LGPL? note it's lgpl and not gpl

File: src/shogun/CMakeLists.txt [hunk 4]
```diff
@@ -400,18 +403,22 @@ ELSE()
...
 ENDIF()
 
 SHOGUN_DEPENDENCIES(
   LIBRARY LpSolve
   SCOPE PRIVATE
-  CONFIG_FLAG USE_LPSOLVE)
+  CONFIG_FLAG USE_LPSOLVE
+  GPL_ONLY)
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: because lpsolve ig LGPL?

File: src/shogun/CMakeLists.txt [hunk 5]
```diff
@@ -400,18 +403,22 @@ ELSE()
...
 
 SHOGUN_DEPENDENCIES(
   LIBRARY ColPack
   SCOPE PRIVATE
-  CONFIG_FLAG HAVE_COLPACK)
+  CONFIG_FLAG HAVE_COLPACK
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: because colpack is LGPL?

File: src/shogun/CMakeLists.txt (hunk 1 above)
Comment (karlnapf) [reply src/shogun/CMakeLists.txt]: it a actually is PD, I got this one wrong

File: src/shogun/CMakeLists.txt (hunk 3 above)
Comment (karlnapf) [reply src/shogun/CMakeLists.txt]: yep you are right this we can link against

File: src/shogun/CMakeLists.txt (hunk 5 above)
Comment (karlnapf) [reply src/shogun/CMakeLists.txt]: will remove

**Comment by karlnapf** [comment]
My motivation was: linking (even dynamically) against GPL code makes the code GPL. Therefore, I added some guards to prevent this. Correct me if I am wrong.
Definitely the LGPL things need to be removed.

File: src/shogun/CMakeLists.txt (hunk 3 above)
Comment (karlnapf) [reply src/shogun/CMakeLists.txt]: I just realised we can also move back the `CNLOPTMinimizer` class to the main repo, which is very good as then all the GP codes are still in the BSD release!

**Comment by karlnapf** [comment]
Apparently the FSF sees dynamic linking as derivative work:
https://en.wikipedia.org/wiki/GNU_General_Public_License#Linking_and_derived_works

**Comment by vigsterkr** [comment]
yeah but in case you use it as a plugin, then it never violates the gpl.

**Comment by vigsterkr** [comment]
so, in this particular case if `src/shogun/lib/Compressor.cpp` would be written in a modular fashion and the lzo decoder would be a plugin then all this can be ignored... for the time being i'd rather suggest to drop LZO as a whole.... as anyways this should not effect anything in the base library  - once modular.

**Comment by vigsterkr** [comment]
and i dont really see anybody complaining of not being able to decompress string features that are lzo compressed... and because of this doxygen comment:
```Note that besides lzo compression, this library is thread safe.```

**Comment by vigsterkr** [comment]
so on the end only GLPK would need this if we call dynamic linking a derivative work, but as soon as there's plugin support this would need to be reverted....

**Comment by karlnapf** [comment]
Sure then let`s leave it

**Pull request closed without merging**